modopt/opt/proximity.py

```diff
@@ -688,3 +689,341 @@ def _cost_method(self, *args, **kwargs):
+        float
+            The alpha value for which sum(theta(alpha[index+1])) >= k
+        float
+            Value of sum(theta(alpha[index]))
+        float
+            Value of sum(theta(alpha[index + 1]))
+        """
+        first_idx = 0
+        data_abs = np.abs(data)
+        last_idx = alpha.shape[0] - 1
+        found = False
+        prev_midpoint = 0
+        cnt = 0  # Avoid infinite looops
+
+        # Checking particular to be sure that the solution is in the array
+        sum_0 = self._compute_theta(data_abs, alpha[0], extra_factor).sum()
+        sum_1 = self._compute_theta(data_abs, alpha[-1], extra_factor).sum()
+        if sum_1 <= self.k_value:
+            midpoint = alpha.shape[0] - 2
+            found = True
+        if sum_0 >= self.k_value:
+            found = True
+            midpoint = 0
+
+        while (first_idx <= last_idx) and not found and (cnt < alpha.shape[0]):
+
+            midpoint = (first_idx + last_idx) // 2
+            cnt += 1
+
+            if prev_midpoint == midpoint:
+
+                # Particular case
+                sum_0 = self._compute_theta(data_abs, alpha[first_idx],
+                                            extra_factor).sum()
+                sum_1 = self._compute_theta(data_abs, alpha[last_idx],
+                                            extra_factor).sum()
+
+                if (np.abs(sum_0 - self.k_value) <= 1e-4):
+                    found = True
+                    midpoint = first_idx
+
+                if (np.abs(sum_1 - self.k_value) <= 1e-4):
+                    found = True
+                    midpoint = last_idx - 1
+                    # -1 because output is index such that
+                    # sum(theta(alpha[index])) <= k
+
+                if (first_idx - last_idx == 2) or (first_idx - last_idx == 1):
+                    sum_0 = self._compute_theta(data_abs, alpha[first_idx],
+                                                extra_factor).sum()
+                    sum_1 = self._compute_theta(data_abs, alpha[last_idx],
+                                                extra_factor).sum()
+                    if (sum_0 <= self.k_value) or (sum_1 >= self.k_value):
+                        found = True
+
+            sum_0 = self._compute_theta(data_abs, alpha[midpoint],
+                                        extra_factor).sum()
+            sum_1 = self._compute_theta(data_abs, alpha[midpoint + 1],
+                                        extra_factor).sum()
+
+            if (sum_0 <= self.k_value) & (sum_1 >= self.k_value):
+                found = True
+
+            elif sum_1 < self.k_value:
+                first_idx = midpoint
+
+            elif sum_0 > self.k_value:
+                last_idx = midpoint
+
+            prev_midpoint = midpoint
+
+        if found:
+            return midpoint, alpha[midpoint], alpha[midpoint + 1], sum_0,\
+                   sum_1
+        else:
+            return -1
```

Comment:
Probably need to raise an error if it has not been found, rather than setting the output to -1, which will cause unhandled error

Reply:
Done